cli/tests/program.rs

```diff
@@ -1350,6 +1356,210 @@ fn test_cli_program_mismatch_buffer_authority() {
+    config.json_rpc_url = test_validator.rpc_url();
+
+    let online_signer = Keypair::new();
+    let online_signer_identity = NullSigner::new(&online_signer.pubkey());
+    let offline_signer = Keypair::new();
+    let buffer_signer = Keypair::new();
+    let buffer_signer_identity = NullSigner::new(&buffer_signer.pubkey());
+    // Typically, keypair for program signer should be different from online signer or
+    // offline signer keypairs.
+    let program_signer = Keypair::new();
+    let program_signer_identity = NullSigner::new(&program_signer.pubkey());
+
+    // Assume fee payer will be either online signer or offline signer (could be completely
+    // separate signer too, but that option is unlikely to be chosen often, so don't bother
+    // testing for it), we want to test both
+    let use_offline_signer_as_fee_payer = rand::thread_rng().gen_bool(0.5);
+
+    config.command = CliCommand::Airdrop {
+        pubkey: None,
+        lamports: 100 * minimum_balance_for_large_buffer, // gotta be enough for this test
+    };
+    config.signers = vec![&online_signer];
+    process_command(&config).unwrap();
+    config.command = CliCommand::Airdrop {
+        pubkey: None,
+        lamports: 100 * minimum_balance_for_large_buffer, // gotta be enough for this test
+    };
+    config.signers = vec![&offline_signer];
+    process_command(&config).unwrap();
+
+    // Deploy upgradeable program with authority set to offline signer
+    config.signers = vec![&online_signer, &offline_signer, &program_signer];
+    config.command = CliCommand::Program(ProgramCliCommand::Deploy {
+        program_location: Some(noop_path.to_str().unwrap().to_string()),
+        fee_payer_signer_index: 0,
+        program_signer_index: Some(2),
+        buffer_signer_index: None,
+        allow_excessive_balance: false,
+        upgrade_authority_signer_index: 1, // must be offline signer for security reasons
+        is_final: false,
+        max_len: Some(max_program_data_len), // allows for larger program size with future upgrades
+        skip_fee_check: false,
+    });
+    config.output_format = OutputFormat::JsonCompact;
+    process_command(&config).unwrap();
+
+    // Prepare buffer to upgrade deployed program to a larger program
+    create_buffer_with_offline_authority(
+        &rpc_client,
+        &noop_large_path,
+        &mut config,
+        &online_signer,
+        &offline_signer,
+        &buffer_signer,
+        &buffer_signer_identity,
+    );
+
+    // Offline sign-only with signature over "wrong" message (with different buffer)
+    config.signers = vec![
+        &offline_signer,
+        &buffer_signer_identity,
+        &program_signer_identity,
+    ];
+    let mut fee_payer_signer_index = 0; // offline signer
+    if !use_offline_signer_as_fee_payer {
+        fee_payer_signer_index = 3; // online signer
+        config.signers.push(&online_signer_identity); // can't (and won't) provide signature, for simplicity
+    }
```

Comment:
micro-nit: it's a bit more rusty to avoid the mut and instead do
```suggestion
    let fee_payer_signer_index = if use_offline_signer_as_fee_payer {
        0
    } else {
        config.signers.push(&online_signer_identity); // can't (and won't) provide signature, for simplicity
        3
    }
```

Reply:
Got it, changed what I found on a recent self-review in https://github.com/solana-labs/solana/pull/33860/commits/194dd375415925d6decefc879e283501278af8e9